File: forms/fertility-assessment/back-end-with-loco/tests/engine/fertility_grader_test.rs
use fertility_assessment_loco_crate::engine::fertility_grader::{
    classify_concern_score, grade, run_rules,
};
use fertility_assessment_loco_crate::engine::fertility_rules::all_rules;
use fertility_assessment_loco_crate::engine::types::*;

#[test]
fn empty_assessment_is_low_concern() {
    let data = AssessmentData::default();
    let result = grade(&data);
    assert_eq!(result.concern_level, "low");
    assert_eq!(result.concern_score, 0);
    assert!(result.fired_rules.is_empty());
}

#[test]
fn classify_concern_score_thresholds() {
    assert_eq!(classify_concern_score(0), "low");
    assert_eq!(classify_concern_score(2), "low");
    assert_eq!(classify_concern_score(3), "moderate");
    assert_eq!(classify_concern_score(6), "moderate");
    assert_eq!(classify_concern_score(7), "high");
    assert_eq!(classify_concern_score(20), "high");
}

#[test]
fn amenorrhoea_fires_cycle_002_rule() {
    let mut data = AssessmentData::default();
    data.menstrual_cycle.cycle_regularity = "absent".to_string();
    let (_score, fired) = run_rules(&data);
    assert!(fired.iter().any(|r| r.id == "FERT-CYC-002"));
}

#[test]
fn low_amh_fires_ovarian_reserve_rule() {
    let mut data = AssessmentData::default();
    data.hormone_profile.amh = Some(4.0);
    let (_score, fired) = run_rules(&data);
    assert!(fired.iter().any(|r| r.id == "FERT-OR-001"));
}

#[test]
fn severe_oligozoospermia_fires_both_rules() {
    let mut data = AssessmentData::default();
    data.partner_semen.semen_analysis_done = "yes".to_string();
    data.partner_semen.semen_concentration_million_per_ml = Some(3.0);
    let (_score, fired) = run_rules(&data);
    // Both FERT-SEM-002 (<16) and FERT-SEM-006 (<5) fire.
    assert!(fired.iter().any(|r| r.id == "FERT-SEM-002"));
    assert!(fired.iter().any(|r| r.id == "FERT-SEM-006"));
}

#[test]
fn abnormal_hsg_fires_tubal_rule() {
    let mut data = AssessmentData::default();
    data.investigations.hysterosalpingogram_done = "yes".to_string();
    data.investigations.hysterosalpingogram_result = "abnormal".to_string();
    let (_score, fired) = run_rules(&data);
    assert!(fired.iter().any(|r| r.id == "FERT-TUB-001"));
}

#[test]
fn pcos_fires_pelvic_rule() {
    let mut data = AssessmentData::default();
    data.medical_surgical_history.polycystic_ovary_syndrome = "yes".to_string();
    let (_score, fired) = run_rules(&data);
    assert!(fired.iter().any(|r| r.id == "FERT-PEL-003"));
}

#[test]
fn rule_ids_are_unique() {
    let rules = all_rules();
    let mut ids: Vec<&str> = rules.iter().map(|r| r.id).collect();
    ids.sort();
    let len_before = ids.len();
    ids.dedup();
    assert_eq!(ids.len(), len_before, "fertility rule IDs must be unique");
}

#[test]
fn high_concern_when_many_high_weight_rules_fire() {
    let mut data = AssessmentData::default();
    // Trigger AMH < 5.4 (weight 3) + amenorrhoea (weight 3) + abnormal HSG (weight 3) = 9 → high.
    data.hormone_profile.amh = Some(3.0);
    data.menstrual_cycle.cycle_regularity = "absent".to_string();
    data.investigations.hysterosalpingogram_done = "yes".to_string();
    data.investigations.hysterosalpingogram_result = "abnormal".to_string();
    let result = grade(&data);
    assert_eq!(result.concern_level, "high");
    assert!(result.concern_score >= 7);
}

File: forms/fertility-assessment/front-end-with-svelte/src/lib/engine/fertility-grader.test.ts
import { describe, it, expect } from 'vitest';
import { calculateConcern, classifyConcernScore } from './fertility-grader';
import { detectAdditionalFlags } from './flagged-issues';
import { fertilityRules } from './rules';
import type { AssessmentData } from './types';

/** A young couple with normal investigations: should score zero and raise no flags. */
function createHealthyPatient(): AssessmentData {
	return {
		demographics: {
			patientFirstName: 'Anna',
			patientLastName: 'Doe',
			patientDateOfBirth: '1996-05-15',
			patientSex: 'female',
			partnerFirstName: 'Ben',
			partnerLastName: 'Doe',
			partnerDateOfBirth: '1994-03-10',
			partnerSex: 'male',
			relationshipDuration: 5,
			ethnicity: ''
		},
		reproductiveHistory: {
			durationTryingMonths: 6,
			priorPregnancies: 0,
			priorLiveBirths: 0,
			priorMiscarriages: 0,
			priorEctopic: 0,
			priorTerminations: 0,
			priorFertilityTreatment: 'no',
			priorTreatmentDetails: '',
			contraceptionStopped: 'yes',
			contraceptionStoppedDate: '2025-01-01'
		},
		menstrualCycle: {
			menarcheAge: 13,
			cycleLengthDays: 28,
			cycleRegularity: 'regular',
			periodDurationDays: 5,
			heavyBleeding: 'no',
			dysmenorrhoea: 'no',
			intermenstrualBleeding: 'no',
			lastMenstrualPeriod: '2026-06-01',
			cycleNotes: ''
		},
		medicalSurgicalHistory: {
			pelvicInflammatoryDisease: 'no',
			endometriosis: 'no',
			polycysticOvarySyndrome: 'no',
			fibroids: 'no',
			thyroidDisorder: 'no',
			diabetes: 'no',
			cancerHistory: 'no',
			cancerTreatmentDetails: '',
			pelvicSurgery: 'no',
			pelvicSurgeryDetails: '',
			sexuallyTransmittedInfections: 'no',
			stiDetails: '',
			otherConditions: ''
		},
		lifestyleFactors: {
			weight: 65,
			height: 168,
			bmi: 23,
			tobaccoStatus: 'never',
			cigarettesPerDay: null,
			alcoholLevel: 'low',
			alcoholUnitsPerWeek: 4,
			caffeineLevel: 'low',
			recreationalDrugs: 'no',
			recreationalDrugDetails: '',
			exerciseFrequency: 'moderate',
			occupationalHazards: 'no',
			occupationalHazardDetails: ''
		},
		medicationsSupplements: {
			currentMedications: [],
			folicAcid: 'yes',
			folicAcidDoseMcg: 400,
			vitaminD: 'yes',
			otherSupplements: ''
		},
		partnerSemen: {
			partnerAgeYears: 32,
			partnerSmoking: 'never',
			partnerAlcohol: 'low',
			partnerOccupationalHazards: '',
			partnerMedicalHistory: '',
			semenAnalysisDone: 'yes',
			semenAnalysisDate: '2026-05-01',
			semenVolumeMl: 3,
			semenConcentrationMillionPerMl: 60,
			semenTotalMotilityPercent: 55,
			semenProgressiveMotilityPercent: 45,
			semenNormalMorphologyPercent: 8,
			semenNotes: ''
		},
		hormoneProfile: {
			fsh: 6,
			lh: 5,
			amh: 25,
			oestradiol: 200,
			tsh: 1.5,
			prolactin: 300,
			testosterone: 1.2,
			progesteroneDay21: 45,
			hormoneTestDate: '2026-05-01',
			hormoneNotes: ''
		},
		investigations: {
			transvaginalUltrasound: 'yes-normal',
			antralFollicleCount: 14,
			hysterosalpingogramDone: 'yes',
			hysterosalpingogramResult: 'normal',
			hysteroscopyDone: 'no',
			hysteroscopyResult: '',
			laparoscopyDone: 'no',
			laparoscopyResult: '',
			otherInvestigations: ''
		},
		clinicalRecommendation: {
			clinicianName: '',
			assessmentDate: '',
			recommendation: '',
			referralUrgency: '',
			additionalNotes: ''
		}
	};
}

describe('Fertility classifyConcernScore', () => {
	it('maps score bands to concern levels', () => {
		expect(classifyConcernScore(0)).toBe('low');
		expect(classifyConcernScore(2)).toBe('low');
		expect(classifyConcernScore(3)).toBe('moderate');
		expect(classifyConcernScore(6)).toBe('moderate');
		expect(classifyConcernScore(7)).toBe('high');
		expect(classifyConcernScore(20)).toBe('high');
	});
});

describe('Fertility Grading Engine', () => {
	it('returns low concern with no fired rules for a healthy young couple', () => {
		const data = createHealthyPatient();
		const result = calculateConcern(data);
		expect(result.concernLevel).toBe('low');
		expect(result.concernScore).toBe(0);
		expect(result.firedRules).toHaveLength(0);
	});

	it('returns moderate concern for irregular cycles plus PCOS', () => {
		const data = createHealthyPatient();
		data.menstrualCycle.cycleRegularity = 'irregular'; // +2
		data.medicalSurgicalHistory.polycysticOvarySyndrome = 'yes'; // +1
		const result = calculateConcern(data);
		expect(result.concernScore).toBe(3);
		expect(result.concernLevel).toBe('moderate');
	});

	it('returns high concern for advanced age plus low ovarian reserve', () => {
		const data = createHealthyPatient();
		data.demographics.patientDateOfBirth = '1983-01-01'; // >= 40 → +2 +2
		data.hormoneProfile.amh = 4.0; // < 5.4 → +3
		const result = calculateConcern(data);
		expect(result.concernScore).toBeGreaterThanOrEqual(7);
		expect(result.concernLevel).toBe('high');
	});

	it('fires the severe male-factor rule for very low concentration', () => {
		const data = createHealthyPatient();
		data.partnerSemen.semenConcentrationMillionPerMl = 3;
		const result = calculateConcern(data);
		expect(result.firedRules.some((r) => r.id === 'FERT-SEM-006')).toBe(true);
		expect(result.firedRules.some((r) => r.id === 'FERT-SEM-002')).toBe(true);
	});

	it('has unique rule IDs', () => {
		const ids = fertilityRules.map((r) => r.id);
		expect(new Set(ids).size).toBe(ids.length);
	});
});

describe('Fertility Flagged Issues Detection', () => {
	it('returns no flags for a healthy young couple', () => {
		const data = createHealthyPatient();
		const flags = detectAdditionalFlags(data);
		expect(flags).toHaveLength(0);
	});

	it('flags advanced maternal age as urgent', () => {
		const data = createHealthyPatient();
		data.demographics.patientDateOfBirth = '1983-01-01';
		const flags = detectAdditionalFlags(data);
		expect(flags.some((f) => f.id === 'FLAG-AGE-001' && f.priority === 'urgent')).toBe(true);
	});

	it('flags missing semen analysis when not yet completed', () => {
		const data = createHealthyPatient();
		data.partnerSemen.semenAnalysisDone = 'no';
		const flags = detectAdditionalFlags(data);
		expect(flags.some((f) => f.id === 'FLAG-SEM-000')).toBe(true);
	});

	it('flags missing pre-conception folic acid', () => {
		const data = createHealthyPatient();
		data.medicationsSupplements.folicAcid = 'no';
		const flags = detectAdditionalFlags(data);
		expect(flags.some((f) => f.id === 'FLAG-SUPP-001')).toBe(true);
	});

	it('sorts flags by priority (urgent first)', () => {
		const data = createHealthyPatient();
		data.demographics.patientDateOfBirth = '1983-01-01'; // urgent
		data.medicationsSupplements.folicAcid = 'no'; // low
		const flags = detectAdditionalFlags(data);
		const order = { urgent: 0, high: 1, medium: 2, low: 3 };
		const priorities = flags.map((f) => order[f.priority]);
		const sorted = [...priorities].sort((a, b) => a - b);
		expect(priorities).toEqual(sorted);
	});
});
